### PR TITLE
BLETelemetry for esp32 devices implementation

### PR DESCRIPTION
**Pull request by flux242**
As long as nobody wanted to finish pull request https://github.com/ExpressLRS/ExpressLRS/pull/2018 I had to do it myself. Basically I implemented all the review comments from PR #2018 on top of the master branch and tested the results.

### CODE REVIEW

**Comment by CapnBry** [comment]
haha I'll look at this after 3.3 releases, but I'm giving it 5 stars for the sweet command line with multiple awks and a perl too 🤩. I was just thinking Saturday about how thank goodness I never see perl code any more and here we are!

**Comment by flux242** [comment]
Well, then I'll continue to torture you millennials with awk and perl one-liners to show you how cool this feature is and what one could do with it.

In the next example I'm converting incoming messages of type 2 (GPS info) into a `json` string. I'm using this `json` format because I have another script which converts such `json` strings into NMEA sentences that are redirected into the `gpsd` daemon

First script to start `gpsd` daemon
```
#!/bin/bash

echo "Creating Virtual Com Port: 0 and 1"

vcpath='/tmp'
socat -d -d pty,link=${vcpath}/virtualcom0,raw,echo=0 pty,link=${vcpath}/virtualcom1,raw,b4800,echo=0 &
socatpid=$!
echo "socat pid=$socatpid"
sleep 2

trap "kill $socatpid &>/dev/null; exit 0" INT TERM EXIT

# if needed disable systemd service first
#systemctl stop gpsd.socket
#systemctl stop gpsd.service

echo "Starting gpsd..."
killall -q gpsd
gpsd -D2 -b -n -N ${vcpath}/virtualcom1
```

Now a one-liner to feed `gpsd` with data
```
gatttool -m 67 -t random -b "5e:08:2d:c5:1a:c8" --char-write-req --handle=0x000d --value=0100 --listen | \
awk -F: '{print $2;fflush()}' | awk '{if($3 == "02"){print;fflush()}}' | \
awk '{printf("{ \"datetime\": \"%s\", \"lat\": %f, \"lon\": %f, \"alt\": %f, \"vel_h\": %f, \"heading\": %f }\n", strftime("%FT%T.000Z",systime()),strtonum("0x"$4$5$6$7)/1e7,strtonum("0x"$8$9$10$11)/1e7,strtonum("0x"$16$17)-1000,strtonum("0x"$12$13)/10,strtonum("0x"$14$15)/100);fflush()}' | \
~/projects/radiosonde/scripts/json2nmea.pl | tee /tmp/virtualcom0
```
(not sure if I calculate alt, vel_h and heading correctly. Haven't seen a docu explaining the data format)

The result printed should look like this (coords are edited):
```
{ "datetime": "2023-07-16T22:28:38.000Z", "lat": 55.850589, "lon": 55.135878, "alt": 0.000000, "vel_h": 0.000000, "heading": 223.900000 }
$GPRMC,222838.000,A,5551.035,N,05508.153,E,0.00,223.90,160723,,*04
$GPGGA,222838.000,5551.035,N,05508.153,E,1,04,0.0,0.000,M,0.0,M,,*64
```

Now any app that connects to `gpsd` can use provided coordinates. This could be for example `foxtrotgps` or an antenna tracker.

Using `tee` multiplexer I could of course split incoming `CRSF` messages to provide with data multiple applications simultaneously. I could provide an example but that much of perl and awk can kill you millennial.

**Comment by arthurson** [comment]
tx16s with internal elrs, using telemetry viewer from googleplay doesn't work , while RomanLut with bluetooth le can finally connected.

**Comment by flux242** [comment]
Some remarks:
1. If you cannot connect again to the `ExpressLRS Telemetry` BLE device again, then try to forget it first
2. If no connection is established within some period of time after transmitter is switched on then TX device will stop advertising. In this case simply deactivate and activate BLE Telemetry in Lua.

I don't like this random addressing myself but I havent found a way to set a specific BSSID address manually. It would be better to get the public address which is used for BT Joystick and increase/decrease the last byte by one

**Comment by EMH3** [comment]
Hello everyone! I flash this pr and it seems to work fine and finally my "abnormal" jjumper aion tx don't freeze & show phantom from a "past life"! Telemetry goes to the phone perfectly and they don't seem to be any bugs yet, although the lua script freezing a little when selecting 500 Hz, but the module itself is fine. Maybe today it will be possible to fly & test in hardcore;) Many thanks to the developers!

**Comment by EMH3** [comment]
Today tested it, so far everything is fine, thanks

**Comment by Pelado-Mat** [comment]
Works fine with my RM Zorro, at least on the bench. It even get a more reliable connection than my BT module with Telemetry Mirror.
Will be trying this branch on a LR flight during the week end.

File: src/lib/BLE/devBLETelemetry.cpp [hunk 1]
```diff
@@ -0,0 +1,301 @@
...
+        ->createCharacteristic(SERIAL_NUMBER_SVC_UUID, NIMBLE_PROPERTY::READ)
+        ->setValue(String(__TIMESTAMP__) + " - " + getMasterUIDString());
+    dInfo
+        ->createCharacteristic(SOFTWARE_NUMBER_SVC_UUID, NIMBLE_PROPERTY::READ)
+        ->setValue(String(version));
+    dInfo
+        ->createCharacteristic(HARDWARE_NUMBER_SVC_UUID, NIMBLE_PROPERTY::READ)
+        ->setValue(String(getRegulatoryDomain()));
+
+    dInfo->start();
```
Comment (RomanLut) [inline src/lib/BLE/devBLETelemetry.cpp:210]:
```suggestion
    dInfo->start();
    
    NimBLEDevice::setPower(ESP_PWR_LVL_P9, ESP_BLE_PWR_TYPE_DEFAULT);
    NimBLEDevice::setPower(ESP_PWR_LVL_P9, ESP_BLE_PWR_TYPE_ADV);
    NimBLEDevice::setPower(ESP_PWR_LVL_P9, ESP_BLE_PWR_TYPE_CONN_HDL0);
    NimBLEDevice::setPower(ESP_PWR_LVL_P9, ESP_BLE_PWR_TYPE_CONN_HDL1);
    NimBLEDevice::setPower(ESP_PWR_LVL_P9, ESP_BLE_PWR_TYPE_CONN_HDL2);
    NimBLEDevice::setPower(ESP_PWR_LVL_P9, ESP_BLE_PWR_TYPE_CONN_HDL3);
    NimBLEDevice::setPower(ESP_PWR_LVL_P9, ESP_BLE_PWR_TYPE_CONN_HDL4);
    NimBLEDevice::setPower(ESP_PWR_LVL_P9, ESP_BLE_PWR_TYPE_CONN_HDL5);
    NimBLEDevice::setPower(ESP_PWR_LVL_P9, ESP_BLE_PWR_TYPE_CONN_HDL6);
    NimBLEDevice::setPower(ESP_PWR_LVL_P9, ESP_BLE_PWR_TYPE_CONN_HDL7);
    NimBLEDevice::setPower(ESP_PWR_LVL_P9, ESP_BLE_PWR_TYPE_CONN_HDL8);    
```

Call at correct place, also use maximum power to ease life of users with poor antennas badly installed inside TX modules...

**Review by RomanLut — CHANGES_REQUESTED** [review]

File: src/lib/BLE/devBLETelemetry.cpp [hunk 2]
```diff
@@ -0,0 +1,301 @@
...
+unsigned short const SOFTWARE_NUMBER_SVC_UUID = 0x2A28;
+unsigned short const HARDWARE_NUMBER_SVC_UUID = 0x2A27;
+unsigned short const MANUFACTURER_NAME_SVC_UUID = 0x2A29;
+
+
+static uint32_t LastTMLLinkStatsPacketMillis = 0;
+static uint32_t LastTLMRCPacketMillis = 0;
+static bool initOnce = false;
+
+static SemaphoreHandle_t _mutex = nullptr;
```
Comment (RomanLut) [inline src/lib/BLE/devBLETelemetry.cpp:43]:
```suggestion
static SemaphoreHandle_t _mutex = nullptr;

class BLEServerCB : public NimBLEServerCallbacks {
  void onConnect(NimBLEServer *pServer, ble_gap_conn_desc  *param) override 
  {
    if ( pServer != nullptr)
    {
        NimBLEDevice::startAdvertising();
    }
  }
} bleServerCB;
```

Continue advertising on connection. This allows connecting multiple devices, .f.e using tablet + ground station + antena tracker.

File: src/lib/BLE/devBLETelemetry.cpp [hunk 3]
```diff
@@ -0,0 +1,301 @@
...
+{
+    if ( pServer != nullptr)
+    {
+        DBGLN("Stopping BLE Telemetry!");
+
+        NimBLEDevice::stopAdvertising();
+        NimBLEDevice::deinit(true);
+        pServer = nullptr;
+        rcCRSF = nullptr;
+        vSemaphoreDelete(_mutex);
```
Comment (RomanLut) [inline src/lib/BLE/devBLETelemetry.cpp:148]:
```suggestion
        pServer->setCallbacks(nullptr);
        pServer = nullptr;
        rcCRSF = nullptr;
        NimBLEDevice::stopAdvertising();
        NimBLEDevice::deinit(true);
        vSemaphoreDelete(_mutex);
```

File: src/lib/BLE/devBLETelemetry.cpp [hunk 4]
```diff
@@ -0,0 +1,301 @@
...
+    }
+}
+
```
Comment (RomanLut) [inline src/lib/BLE/devBLETelemetry.cpp:151]:
```suggestion

static void assignRandomAddress()
 {
    uint8_t blead[6];
    memcpy( blead, NimBLEDevice::getAddress().getNative(), 6 );
    blead[5] |= 0xC0;  //random address should have 11 at most significat bits
    blead[0]++;
    NimBLEDevice::setOwnAddrType(BLE_OWN_ADDR_RANDOM);
    ble_hs_id_set_rnd(blead);
 }

 static void onAdvertisingStopped(NimBLEAdvertising *pAdv)
 {
    if (pServer != nullptr)
    {
        assignRandomAddress();
        pAdv->start(0, &onAdvertisingStopped);
    }
 }
 
```

Add address assignment method and advertisement callback

File: src/lib/BLE/devBLETelemetry.cpp [hunk 5]
```diff
@@ -0,0 +1,301 @@
...
+    }
+    NimBLEDevice::setOwnAddrType(BLE_OWN_ADDR_RANDOM);
+
+    //Set MTU to max packet length + 3 bytes to be able to send packets longer then default 20 bytes.
+    //Should be set on both ends - smaller from two is used.
+    BLEDevice::setMTU(CRSF_MAX_PACKET_LEN + 3);
+
+    /** Set low transmit power, default is 6db */
+    BLEDevice::setPower(ESP_PWR_LVL_P6);
+    pServer = NimBLEDevice::createServer();
```
Comment (RomanLut) [inline src/lib/BLE/devBLETelemetry.cpp:188]:
```suggestion
    pServer = NimBLEDevice::createServer();
    pServer->setCallbacks(&bleServerCB, true);
```

File: src/lib/BLE/devBLETelemetry.cpp [hunk 6]
```diff
@@ -0,0 +1,301 @@
...
+        ->createCharacteristic(HARDWARE_NUMBER_SVC_UUID, NIMBLE_PROPERTY::READ)
+        ->setValue(String(getRegulatoryDomain()));
+
+    dInfo->start();
+
+    NimBLEAdvertising *pAdvertising = NimBLEDevice::getAdvertising();
+    pAdvertising->addServiceUUID(rcService->getUUID());
+
+    pAdvertising->addServiceUUID(dInfo->getUUID());
+    pAdvertising->start();
```
Comment (RomanLut) [inline src/lib/BLE/devBLETelemetry.cpp:216]:
```suggestion
    pAdvertising->start(0, &onAdvertisingStopped);
```
Regardles specified timeout value, BLE library stops advertising after ~10 minutes. Add callback and restart  advertising when stopped.

**Comment by RomanLut** [comment]
Please accept proposed changes. They solve the following problems:
- low BLE transmission power
- advertisiement stops after ~10 minutes if no clients are connected
- implemented: service address is generated from public address, so it does not change every time
- implemented: allows connecting multiple devices

File: src/lib/BLE/devBLETelemetry.cpp (hunk 1 above)
Comment (flux242) [reply src/lib/BLE/devBLETelemetry.cpp:210]: shouldnt setting only ESP_BLE_PWR_TYPE_DEFAULT be enough? According to the docu default   is used for all other modes that were not set explicitly

Comment (RomanLut) [reply src/lib/BLE/devBLETelemetry.cpp:210]: Setting default only does not have effect.
I checked RSSI with this app:
https://play.google.com/store/apps/details?id=com.macdom.ble.blescanner&hl=en

RSSI improves if all settings are set.

**Comment by flux242** [comment]
I've just pushed updated code based on @RomanLut comments. There're some parts that I changed though:
- assigning 'constant' address code didn't work, so I kept the old implementation
- power level set to 4mW instead of 8mW. 4mW should be enough imo and I also do not want my transmitter battery drain faster for nothing

Multiple clients connection tested and it works

File: src/lib/BLE/devBLETelemetry.cpp (hunk 2 above)
Comment (flux242) [reply src/lib/BLE/devBLETelemetry.cpp:42]: added

File: src/lib/BLE/devBLETelemetry.cpp (hunk 4 above)
Comment (flux242) [reply src/lib/BLE/devBLETelemetry.cpp:162]: that code doesn't work. Still a new address generated every time. So old code is used

File: src/lib/BLE/devBLETelemetry.cpp (hunk 5 above)
Comment (flux242) [reply src/lib/BLE/devBLETelemetry.cpp:211]: fixed

File: src/lib/BLE/devBLETelemetry.cpp (hunk 1 above)
Comment (flux242) [reply src/lib/BLE/devBLETelemetry.cpp:234]: power level set to 4mW instead

**Comment by Rodrigo-NH** [comment]
Flashed to NamimnoRC 2.4 OLED RX and BetaFpv Nano TX. It works, can see telemetry in Telemetry View. Pitch is inverted in artificial horizon though (using ardupilot), must be something else.

**Comment by AlanAlger001** [comment]
If used in gemini mode, the hardware will keep restarting.

**Comment by flux242** [comment]
> If used in gemini mode, the hardware will keep restarting.

yet another point to not use gemini

**Comment by howels** [comment]
> > If used in gemini mode, the hardware will keep restarting.
> 
> yet another point to not use gemini

Not really a fair comment. This BT telem PR takes cycles away from the core purpose of the eLRS code - to process packets - in order to route telemetry conveniently via BT. This PR was already known to cause issues with faster packet rates, which was not the fault of the rest of the eLRS codebase. Gemini is processing more packets than traditionally uni-radio TX processing so not surprising that spare CPU cycles are lower with Gemini that other modes.

**Comment by flux242** [comment]
> This PR was already known to cause issues with faster packet rates, 

your statement is based on tests you performed or on some rumors? If latter is the case please perform the test and then come back again

gemini uses 2 times the bandwidth for no apparent reason which means - less bandwidth for others

**Comment by howels** [comment]
> > This PR was already known to cause issues with faster packet rates,
> 
> your statement is based on tests you performed or on some rumors? If latter is the case then please perform the test and then come back again
> 
> gemini uses 2 times the bandwidth for no apparent reason which means - less bandwidth for others

My friend reviewed the system, Gemini works very well. If you want to throw shade on the project then perhaps go onto facebook - the Github repo is not the correct place for this.

**Comment by flux242** [comment]
> My friend reviewed the system

right, feel free to tell us what your mum thinks about it too

**Comment by AlanAlger001** [comment]
Let me talk about the hardware I use, I am a gemini DIY, the main control esp32 pico d4 has only 4m flash, I added a spi screen, i2c gyroscope, pwm fan, it may be caused by too many peripheral devices, so Next, I plan to use esp32 s3 as the main controller for testing. Compared with esp32, esp32s3 has more IO ports and can use multiple IOs for memory expansion. I have tested the current hardware for a long time, including stability testing. We Everyone wants to promote the ELRS project, so I just share the problems I encountered with everyone and seek solutions.

**Comment by AlanAlger001** [comment]
> 先说一下我使用的硬件，我是双子DIY，主控esp32 pico d4只有4m flash，我加了spi屏，i2c陀螺仪，pwm风扇，可能是外围设备太多造成的，所以下一篇，我打算使用esp32 s3作为主控制器进行测试。与esp32相比，esp32s3拥有更多的IO端口，可以使用多个IO进行内存扩展。我对现在的硬件进行了很长时间的测试，包括稳定性测试。我们大家都想推广ELRS项目，所以我只是把我遇到的问题分享给大家，寻求解决方案。

I have found that the reason for hardware restart in gemini mode is a problem caused by the gyroscope. After I commented out the gyroscope configuration, it can run normally. The problem may be that the gyroscope takes up too many resources, or it may be a conflict. , you also need to check the code analysis

**Comment by konstkarapan** [comment]
I've been using this PR for quite some time and I have to say, I love it. I noticed, however, that sometimes my phone randomly disconnects and I am not able to reconnect again for a couple of minutes. My tx is rm zorro with int elrs. This happened 2-3 times but since I was flying with goggles I couldn't really investigate it. I will try to, but I felt I should let you guys know.
Tell me if you need more info about my hardware👍

**Comment by flux242** [comment]
>  Tell me if you need more info about my hardware

what update rate are you using? Ive tested it up to 500Hz 1:2 and it works stable.

**Comment by konstkarapan** [comment]
> >  Tell me if you need more info about my hardware
> 
> what update rate are you using? Ive tested it up to 500Hz 1:2 and it works stable.

I use 100hz, 1:8

**Comment by flux242** [comment]
> I use 100hz, 1:8

this should be ok. Not sure what causes the problem. Have you tried to reboot the radio in such a situation ?

**Comment by flux242** [comment]
>  Does it cause problem with some other ble client?

it never worked for me even with the telemetry app

**Comment by GT-Volk** [comment]
Thank you! This is one of the most useful features for me! After switching to the crummy radio Jumper T20, I really missed being able to display telemetry on my phone! This merge just saved me) I’ve been using it successfully for a couple of months now, on 250hz drones and 100hz airplanes.
Is there anything I can do to speed up the merge of this request into the main branch? This is really a very useful feature!

**Comment by Rodrigo-NH** [comment]
> I've been using this PR for quite some time and I have to say, I love it. I noticed, however, that sometimes my phone randomly disconnects and I am not able to reconnect again for a couple of minutes. My tx is rm zorro with int elrs. This happened 2-3 times but since I was flying with goggles I couldn't really investigate it. I will try to, but I felt I should let you guys know. Tell me if you need more info about my hardware👍

When my phone lose connection I need to go BL settings in phone, scan for nearby devices. Then a second BLETelemetry device appears with the same same. Then I need to connect to this 'new' device to have it working again.

**Comment by konstkarapan** [comment]
> > I've been using this PR for quite some time and I have to say, I love it. I noticed, however, that sometimes my phone randomly disconnects and I am not able to reconnect again for a couple of minutes. My tx is rm zorro with int elrs. This happened 2-3 times but since I was flying with goggles I couldn't really investigate it. I will try to, but I felt I should let you guys know. Tell me if you need more info about my hardware👍
> 
> When my phone lose connection I need to go BL settings in phone, scan for nearby devices. Then a second BLETelemetry device appears with the same name (repeated). Then I need to connect to this 'new' device to have it working again.

There's a high chance the same thing happens to me. I haven't been able to confirm it yet since it always happens when I am flying (with goggles on), but I will try to recreate this on the bench when I have some free time.

**Comment by flux242** [comment]
I was just thinking Saturday about how thank goodness I never see telemetry in my action cam footages! So, I've written an awk script that parses telemetry csv file created by RomanLut's app and generate an overlay. And here we are!
www.youtube.com/watch?v=HDlKoyEhcxg

**Comment by konstkarapan** [comment]
> I was just thinking Saturday about how thank goodness I never see telemetry in my action cam footages! So, I've written an awk script that parses telemetry csv file created by RomanLut's app and generates an overlay. And here we are!
> www.youtube.com/watch?v=HDlKoyEhcxg

Cool, can you share the script?

**Comment by konstkarapan** [comment]
> > > I use 100hz, 1:8
> > 
> > this should be ok. Not sure what causes the problem. Have you tried to reboot the radio in such a situation ?
> 
> It happened mid flight so I couldn't, but after 2-3 minutes I was able to reconnect again

Since updating the app on the phone I didn't have any disconnects, so the problem solved 👍

**Comment by flux242** [comment]
> Since updating the app on the phone I didn't have any disconnects, so the problem solved 👍
I'm not sure about that. I had a disconnect too once. Solved by restarting the radio. I might look into this again next year

**Comment by flux242** [comment]
so, Ive investigated a bit deeper and I could reproduce the "disconnect" behavior. After some connect/disconnect repeats it would not connect anymore and then the "ExpressLRS Telemetry" device would appear but with a different MAC address which shouldn't happen because it is supposed to be fixed. I'm not sure why this happens, maybe a bug in the BLE stack. So, if this happens the solution currently is to disable/enable BLE Telemetry in the lua script.

**Comment by flux242** [comment]
The latest change might solve that "disconnect" problem. At least I couldn't reproduce it by repeatedly connecting/disconnecting the radio.

**Comment by kasatka60** [comment]
How to build the firmware?

**Comment by kvalexey** [comment]
Hello.
Is it planned to upgrade project to elrs configurator 1.6.1? 
ELRS configurator 1.6.0 bild success, but in 1.6.1 error occured.
Some new devices(pocket,bandit) are only in 1.6.1

**Comment by flux242** [comment]
> Hello. Is it planned to upgrade project to elrs configurator 1.6.1? 

should work now

**Comment by kvalexey** [comment]
> > Hello. Is it planned to upgrade project to elrs configurator 1.6.1?
> 
> should work now

Sank you. It work now

**Comment by CopterFail** [comment]
> > Hello. Is it planned to upgrade project to elrs configurator 1.6.1? 
> 
> should work now
> 
> 

Works well 👍 thank you
Does anyone know sample esp32 code to receive the ble telemetry?

**Comment by flux242** [comment]
I have fixed a bug that has prevented proper telemetry update. If you're still using this patch please recompile the FW with the latest commit.

**Comment by kvalexey** [comment]
Hello. I create bin file with this PR(in main page selected 3.3.2 version). When i try to flash it throw WIFI i see error: Could Not Activate this firmware
If i select at main page 3.3.1 and generate bin file for PR the result file flash throw wifi correctly
Some of my TX occured only in 3.3.2 version.
TX with problem: RM 2.4 Ranger, RM Bandit 900 micro

**Comment by kasatka60** [comment]
What if you try to flash it via USB?

**Comment by kvalexey** [comment]
> What if you try to flash it via USB?

Yes. Throw usb flashing was sucessful

**Comment by CopterFail** [comment]
My ESP32 cannot receive "CRSF_FRAMETYPE_BARO_ALTITUDE = 0x09" telegrams , but the baro value is displayed in the telemetry of the transmitter. All the other CRSF frame type work well.
The maximum distance before loosing the BLE connection is 1m. Any hints to increase this?

**Comment by CopterFail** [comment]
The BLETelemetry contains 12 rc-channels, but 16 were used.

**Comment by CopterFail** [comment]
My Antenna Tracking Project ist based ON this Pull requests. I wunder if , or when this pr become part of the Project? I think it will not stay forever as pr?

**Comment by konstkarapan** [comment]
I know it causes some issues when being used with higher packet rates. Why don't we just disable this feature for these higher rates and call it a day? I don't think racers of freestylers will mind. Or are there any other problems that need to be solved? I've been using this pr for several months and except from the disconnect problem that has been solved it works flawlessly for me. Is there anything we can do to speed up the integration? It's definitely a very useful tool.

**Comment by konstkarapan** [comment]
> > and except from the disconnect problem that has been solved 
> 
> I've started using it recently with the T20, and facing disconnects after ~10 min, but I can't reproduce it on the table, only during flight. Packet rate is 150Hz. Maybe it's of high power of TX? BLE turns off completely, and can be found again only after switching it off/on in the script.

In my case it turned out to be client related. I was using romam lut's telemetry viewer app and updating it to the latest version solved the issue, at least for me. Since then I didn't have any disconnects even at 30min+ flights

**Comment by GT-Volk** [comment]
@flux242 Please update the pull request to version 3.4, thanks!

**Comment by aszczu** [comment]
Loaded on Radiomaster Ranger Micro 2.4Ghz-  Neither my phone or PC can't finish pairing process. ExpressLRS Telemetry bluetooth device is properly detected, but fail to connect. Restarting the module or bluetooth, and bluetooth function change nothing.

**Comment by aszczu** [comment]
> > Loaded on Radiomaster Ranger Micro 2.4Ghz-  Neither my phone or PC can't finish pairing process. ExpressLRS Telemetry bluetooth device is properly detected, but fail to connect. Restarting the module or bluetooth, and bluetooth function change nothing.
> 
> What program do you use at the client side?

I wanted to use Android telemetry app by RomanLut, but I didn't went that far- I fail at connecting to  bluetooh davice, I don't think that application have anything to do with it.

**Comment by konstkarapan** [comment]
> > > Loaded on Radiomaster Ranger Micro 2.4Ghz-  Neither my phone or PC can't finish pairing process. ExpressLRS Telemetry bluetooth device is properly detected, but fail to connect. Restarting the module or bluetooth, and bluetooth function change nothing.
> > 
> > What program do you use at the client side?
> 
> I wanted to use Android telemetry app by RomanLut, but I didn't went that far- I fail at connecting to  bluetooh davice, I don't think that application have anything to do with it.
> 

No need to pair, just connect through the app

**Comment by flux242** [comment]
use 2731

**Pull request closed without merging**